fractional/ligo/src/fractional_dao.mligo

```diff
@@ -0,0 +1,296 @@
+      }
+  else
+    let tx_op = make_transfer p.vote in
+    let new_s = clean_after_transfer (p.vote, ownership.ownership_token, s) in
+    [tx_op], s
+  
+let flush_expired (vote, voting_period, pending_votes
+    : transfer_vote * nat * pending_votes) : pending_votes =
+  match Big_map.find_opt vote pending_votes with
+  | None -> (failwith "VOTE_DOES_NOT_EXIST" : pending_votes)
+  | Some info ->
+    if Tezos.now - info.timestamp > int(voting_period)
+    then Big_map.remove vote pending_votes
+    else (failwith "VOTE_NOT_EXPIRED" : pending_votes)
+
+let get_ownership_token (p, owned_nfts : ownership_token_param * ownership) : operation =
+  match Big_map.find_opt p.nft_token owned_nfts with
+  | None -> (failwith "NO_OWNERSHIP" : operation)
+  | Some ownership -> 
+    Tezos.transaction ownership.ownership_token 0mutez p.callback
+
+type dao_entrypoints =
+  | Fa2 of fa2_entry_points (* handling ownership FA2 fungible tokens *)
+  | Set_ownership of set_ownership_param
+  | Vote_transfer of vote_transfer_param
+  | Flush_expired of transfer_vote
+  | Admin of simple_admin
+  | Ownership_token of ownership_token_param
+
+let dao_main (p, s : dao_entrypoints * dao_storage) : operation list * dao_storage =
+  match p with
+  | Fa2 fa2 -> 
+    let u = fail_if_paused s.admin in
+    let ops, new_ownership = fa2_main(fa2, s.ownership_tokens) in
+    ops, { s with ownership_tokens = new_ownership; }
+
+  | Set_ownership op ->
+    let u = fail_if_not_admin s.admin in
+    let new_s = set_ownership(op, s) in
+    ([] : operation list), new_s
+
+  | Ownership_token p ->
+    let op = get_ownership_token (p, s.owned_nfts) in
+    [op], s
+
+  | Vote_transfer vp ->
+    let u = fail_if_paused s.admin in
+    let ops, new_s = vote_transfer(vp, s) in
+    ops, new_s
+
+  | Flush_expired vp ->
+    let u = fail_if_paused s.admin in
+    let ownership = match Big_map.find_opt vp.nft_token s.owned_nfts with
+    | None -> (failwith "NO_OWNERSHIP" : nft_ownership)
+    | Some o -> o
+    in
+    let new_pending = flush_expired (vp, ownership.voting_period, s.pending_votes) in
+    ([] : operation list), { s with pending_votes = new_pending; }
+
+  | Admin ap -> 
+    let ops, new_admin = simple_admin (ap, s.admin) in
+    ops, { s with admin = new_admin; }
+
+
+let sample_storage : dao_storage = {
+  ownership_tokens = {
+    ledger = (Big_map.empty : ledger);
+    operators = (Big_map.empty : operator_storage);
+    token_total_supply = (Big_map.empty : token_total_supply);
+    token_metadata = (Big_map.empty : token_metadata_storage);
+  };
+  next_ownership_token_id = 0n;
+  admin  = {
+    admin = ("tz1YPSCGWXwBdTncK2aCctSZAXWvGsGwVJqU" : address);
+    pending_admin = (None : address option);
+    paused = false;
+  };
+  vote_nonce = 0n;
+  owned_nfts = (Big_map.empty : ownership);
+  pending_votes = (Big_map.empty : pending_votes);
+  metadata  = Big_map.literal [
+    ("", Bytes.pack "tezos-storage:content" );
```

Comment:
I guess this is actually never used, but I don't think it's good to confuse people by using `pack`.

Reply:
Oh, I actually use it to run ligo off-chain `compile-storage` command to generate Michelson for contract origination.